src/js/components/FileInput/FileInput.js

```diff
@@ -239,8 +238,32 @@ const FileInput = forwardRef(
         nextFiles.splice(index, 1);
       }
       setFiles(nextFiles);
-      if (onChange) onChange(event, { files: nextFiles });
-      if (nextFiles.length === 0) inputRef.current.value = '';
+
+      // Need to have a way to track the files other than an array
+      // since inputRef.current.files is a read-only FileList
+      // https://stackoverflow.com/a/64019766
+      /* eslint-disable no-undef */
+      const dt = new DataTransfer();
+      const curFiles = inputRef.current.files;
+      if (index === 'all' || nextFiles.length === 0)
+        inputRef.current.value = '';
+      for (let i = 0; i < curFiles.length; i += 1) {
+        const curfile = curFiles[i];
+        if (index !== i) dt.items.add(curfile);
+      }
+
+      if (onChange) {
+        // We are using dispatchEvent to align the onChange event from
+        // the input, instead of the onClick event from the remove button.
+        const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
+          window.HTMLInputElement.prototype,
+          'files',
+        ).set;
+        nativeInputValueSetter.call(inputRef.current, dt.files);
```

Comment:
Shouldn't we be doing this irregardless of whether `onChange` is provided? That way a test program could introspect the DOM to get the current value.

Reply:
i can move this out of the if statement based on `onChange`